### PR TITLE
fix(core): ignore strategies do not match directories

### PR DESCRIPTION
**Pull request by danilobuerger**
ignore strategies like git, expect directories to be passed in with a trailing slash in order to properly match.

consider a source tree:

```
├── hello
│   └── index.html
├── index.html
```

now we want to exclude everything except all html files (including all subdirectories).

this can be done with a gitignore of:
```
*
!*/
!*.html
```

in order for the git ignore strategy to properly match this, it needs to know that hello is a directory.

this patch just does that by appending a trailing slash for the ignore strategies.

Fixes https://github.com/aws/aws-cdk/issues/9146

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by danilobuerger** [comment]
The build failure seems unrelated. At least I can't find the cause from the logs thats related to my code change.

File: packages/@aws-cdk/core/lib/fs/copy.ts [hunk 1]
```diff
@@ -15,15 +15,16 @@ export function copyDirectory(srcDir: string, destDir: string, options: CopyOpti
...
   }
 
-  const files = fs.readdirSync(srcDir);
+  const files = fs.readdirSync(srcDir, { withFileTypes: true });
   for (const file of files) {
-    const sourceFilePath = path.join(srcDir, file);
+    const sourceFilePath = path.join(srcDir, file.name);
 
-    if (ignoreStrategy.ignores(sourceFilePath)) {
+    const ignorePath = sourceFilePath + (file.name && file.isDirectory() ? path.sep : "");
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/core/lib/fs/copy.ts:22]: Build failure message:
```
@aws-cdk/core: /codebuild/output/src689345531/src/github.com/aws/aws-cdk/packages/@aws-cdk/core/lib/fs/copy.ts
@aws-cdk/core:   22:87  error  Strings must use singlequote  quotes
```

File: packages/@aws-cdk/core/lib/fs/ignore.ts [hunk 2]
```diff
@@ -117,6 +117,9 @@ export class GlobIgnoreStrategy extends IgnoreStrategy {
     }
 
     let relativePath = path.relative(this.absoluteRootPath, absoluteFilePath);
+    if (relativePath) {
+      relativePath += absoluteFilePath.endsWith(path.sep) ? path.sep : "";
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/core/lib/fs/ignore.ts:121]: Build failure message:
```
@aws-cdk/core: /codebuild/output/src689345531/src/github.com/aws/aws-cdk/packages/@aws-cdk/core/lib/fs/ignore.ts
@aws-cdk/core:   121:72  error  Strings must use singlequote  quotes
@aws-cdk/core:   181:72  error  Strings must use singlequote  quotes
@aws-cdk/core:   227:72  error  Strings must use singlequote  quotes
```

**Review by TheRealAmazonKendra — DISMISSED** [review]

**Comment by TheRealAmazonKendra** [comment]
A few comments: Firstly, please take another look at the build failures, they're directly related to this code change.

Additionally, please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request)).

Lastly, PRs with changes that are not statically guaranteed by the type system must have tests to make sure that (a) the new behavior works as intended and (b) it is not accidentally undone by future changes (See [Contributing Guide, Work Your Magic](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-3-work-your-magic)).

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))

PRs with changes that are not statically guaranteed by the type system must have tests to make sure that (a) the new behavior works as intended and (b) it is not accidentally undone by future changes (See [Contributing Guide, Work Your Magic](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-3-work-your-magic) (https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-3-work-your-magic)).

**Comment by danilobuerger** [comment]
Hi @TheRealAmazonKendra, thanks for your helpful feedback! I really appreciate it. Sorry I didn't see the failures in the build log. I will try to get some meaningful tests in.

**Comment by danilobuerger** [comment]
@TheRealAmazonKendra it looks like its all green now. Thanks again for all your help on getting me there!

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Pull request closed without merging**